### DISCUSSION TITLE
BR2 lifting operators only for boundary elements

### DISCUSSION THREAD

**Discussion by hughcars** [mfem/mfem#3246] [Ideas]
Weakly imposing boundary conditions for FEM can give super convergence in output functionals, coming from the adjoint consistency. To do this, one approach is the SIP/Nitsche penalty style face integrals. These work, but then the complexity becomes the choice of the penalization parameter, which is ad-hoc at best.

An alternative is Bassi and Rebay's second scheme, which computes lifting operators that are used to weight the jump term. The stability criterion then becomes eta >= nfaces, which can be easily verified. Lifting operators can also be used on boundary elements, and make for excellent weak boundary condition enforcement. The extra upside is they can be computed for the cost of an elemental mass matrix inversion per element, and given their linearity, statically condensed so as to never appear in the global solve.

To introduce them in mfem, would require the ability to define a volume field, defined only in elements attached to boundaries, and with only one vector field per boundary face. Alternatively, these variables can never be computed and only used via the static elimination form. This would then mean construction of an operator that incorporates an elemental mass matrix inversion.

**Comment by pazner** [comment]
Hi @hughcars, perhaps `DGDiffusionBR2Integrator` will be close to what you need? You can add this integrator only to the boundary faces using `BilinearForm::AddBdrFaceIntegrator`.
